### PR TITLE
Introduce territories, refs etalab/udata#284

### PR DESCRIPTION
**Pull request by davidbgk**

### CODE REVIEW

File: udata_gouvfr/templates/territory.html [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            title="{{ _('I belong to this city') }}">
+            <span class="glyphicon glyphicon-user"></span>
+            I’m legitimate<br/>
+            to rule this city
+        </button>
+        -->
+        <br>
+    </aside>
+</div>
+<br/>
```
Comment (vinyll) [inline udata_gouvfr/templates/territory.html]: ?

File: udata_gouvfr/templates/territory.html [hunk 2]
```diff
@@ -0,0 +1,71 @@
+{% extends theme("layouts/1-column.html") %}
+
+{% block main_content %}
+{% set logo=territory.logo() %}
+<div class="row">
+    <div class="col-sm-9">
```
Comment (vinyll) [inline udata_gouvfr/templates/territory.html]: shouldn't this be an `<article>`?

File: udata_gouvfr/commands.py [hunk 3]
```diff
@@ -18,21 +24,46 @@
...
-                log.info(u'{title} not found'.format(title=title))
-            else:
-                dataset.badges.append(badge)
-                dataset.save()
+def fetch_territories_logos():
+    """Retrieves images of flags or blazons from wikimedia."""
+    log.info('As of January 2016, the download is about 1GB. Time to relax.')
+    # A bit tricky but otherwise you cannot guess the final file URL.
+    DBPEDIA_MEDIA_URL = 'http://commons.wikimedia.org/wiki/Special:FilePath/'
+    LOGOS_FOLDER_PATH = os.path.join(current_app.config['FS_ROOT'], 'logos')
```
Comment (noirbizarre) [inline udata_gouvfr/commands.py]: You can use the storage directly: `LOGOS_FOLDER_PATH = logos.root`

File: udata_gouvfr/commands.py [hunk 4]
```diff
@@ -18,21 +24,46 @@
...
+                for chunk in r.iter_content(chunk_size=1024):
+                    file_destination.write(chunk)
+
+    log.info('Done')
+
+
+@m.command
+def collect_references_files():
+    """Retrieve locally CSV files in use for dynamic territories' resources."""
+    REFERENCES_PATH = os.path.join(current_app.config['FS_ROOT'], 'references')
```
Comment (noirbizarre) [inline udata_gouvfr/commands.py]: Same as logos

File: udata_gouvfr/views.py [hunk 5]
```diff
@@ -158,6 +166,67 @@ def terms():
...
+    """
+    args = multi_to_dict(request.args)
+    if 'territory_attr' not in args or 'csv_column' not in args:
+        return abort(404)
+    for resource in dataset.resources:
+        if resource.id == resource_id:
+            break
+
+    file_name = resource.url.split('/')[-1]
+    resource_path = os.path.join(
```
Comment (noirbizarre) [inline udata_gouvfr/views.py]: `resource_path = references.path(file_name)`

File: udata_gouvfr/views.py [hunk 6]
```diff
@@ -158,6 +166,67 @@ def terms():
...
+        current_app.config['FS_ROOT'],
+        'references',
+        file_name)
+    match = getattr(territory, args['territory_attr']).encode('utf-8')
+
+    csvfile_out = StringIO.StringIO()
```
Comment (noirbizarre) [inline udata_gouvfr/views.py]: Why not reuse the `udata.frontend.csv` module ?

File: udata_gouvfr/models.py [hunk 7]
```diff
@@ -28,3 +31,210 @@
...
+
+class ResourceBasedTerritoryDataset(TerritoryDataset):
+    dataset_id = ''
+    resource_id = ''
+    territory_attr = ''
+    csv_column = ''
+
+    @property
+    def url(self):
+        return '{url}?{parameters}'.format(
```
Comment (noirbizarre) [inline udata_gouvfr/models.py]: No need to anually concatenate the parameters, Flask automatically add (and encode) them if not defined in the url pattern:
```python
return url_for('gouvfr.territory_dataset_resource', territory=self.territory,
                                                    dataset=self.dataset_id,
                                                    resource_id=self.resource_id,
                                                    territory_attr=self.territory_attr,
                                                    csv_column=self.csv_column)
```

File: udata_gouvfr/models.py [hunk 8]
```diff
@@ -28,3 +31,210 @@
...
+            }),
+            parameters=urllib.urlencode({
+                'territory_attr': self.territory_attr,
+                'csv_column': self.csv_column
+            })
+        )
+
+    @property
+    def external_url(self):
+        return '{server}{path}'.format(server=self.server, path=self.url)
```
Comment (noirbizarre) [inline udata_gouvfr/models.py]:
```python
return url_for('gouvfr.territory_dataset_resource', territory=self.territory,
                                                    dataset=self.dataset_id,
                                                    resource_id=self.resource_id,
                                                    territory_attr=self.territory_attr,
                                                    csv_column=self.csv_column,
                                                    _external=True)

File: udata_gouvfr/models.py [hunk 9]
```diff
@@ -28,3 +31,210 @@
...
+
+    @property
+    def server(self):
+        return '{scheme}://{domain}'.format(
+            scheme=current_app.config['USE_SSL'] and 'https' or 'http',
+            domain=current_app.config['SERVER_NAME'])
+
+    @property
+    def api_url(self):
+        return '{server}/api/1/territories/{path}/'.format(
```
Comment (noirbizarre) [inline udata_gouvfr/models.py]: Can't find the API definition but you should use the `url_for` helper:
```python
url_for('api.territory', territory=self.slug)
```

Comment (davidbgk) [inline udata_gouvfr/models.py]: There is no `api.territory` url.

File: udata_gouvfr/commands.py (hunk 3 above)
Comment (davidbgk) [inline udata_gouvfr/commands.py]: Oh nice :)

File: udata_gouvfr/models.py (hunk 7 above)
Comment (davidbgk) [inline udata_gouvfr/models.py]: :kissing:

File: udata_gouvfr/commands.py (hunk 3 above)
Comment (davidbgk) [inline udata_gouvfr/commands.py]: Wait, where is `logos` coming from?

**Pull request merged**